src/analyzer/analyzersilence.cpp

```diff
@@ -30,13 +52,29 @@ Iterator first_sound(Iterator begin, Iterator end) {
     });
 }
 
+template<typename ForwardIterator>
+ForwardIterator last_fade_in_sound(ForwardIterator begin, ForwardIterator end) {
+    return std::find_if(begin, end, [](const auto elem) {
+        return fabs(elem) >= kFadeInThreshold;
+    });
+}
+
+template<typename Iterator>
+Iterator first_fade_out_sound(Iterator begin, Iterator end) {
+    return std::find_if(begin, end, [](const auto elem) {
+        return fabs(elem) >= kFadeOutThreshold;
+    });
+}
+
```

Comment:
this is a copy of the above for very little reason. Consider deduplicating by simply taking the Threshold as a parameter.

Reply:
The second one goes backward, while the first is strictly forward.  I spent a long time playing around with these, trying to determine why I was getting weird results (and so I was keen to keep them separate).  It turns out AutoDj processing can take place before a track is fully analysed, so I am less invested in that now.

Reply:
The functions are identical. In this context `ForwardIterator` and `Iterator` are just names that resolve to the underlying iterator implementation when called. So there is no difference in terms of what function can be called when. A [`std::reverse_iterator`](https://en.cppreference.com/w/cpp/iterator/reverse_iterator) is a ["bidirectional iterator"](https://en.cppreference.com/w/cpp/named_req/BidirectionalIterator) and thus also a ["forward iterator"](https://en.cppreference.com/w/cpp/named_req/ForwardIterator).

Reply:
`std::find_if` only requires a forward iterator.

Reply:
If you like I'll refactor all of those into one, including the db60.

Reply:
yes, that would be nice. Thank you.

Reply:
As I went to do this I've had second thoughts.

Can you look again at this, and consider how much if any savings will be made by a refactor here?  Since are are assigning parameters to the stack in a call for functions have are already present in std lib (so nothing additional) then I don't see a saving, in fact, we are adding to the time to complete.  And, the code is more readable with several functions (note the difference is a const).  Unnecessarily concise code does not equal readability, right?

Reply:
well, the think that trips me up when reading is that the body is repeated several times. Its difficult to spot where the difference is. I don't expect any meaningful performance improvements from this. If you want to give the difference functions a name, you're welcome to do so, but remove the duplicated bodies.
Something this would work for me:

```cpp
template<typename It, typename T>
It find_first_above_threshold(It begin, It end, T threshold) {
    return std::find_if(begin, end, [threshold](const auto& elem) {
        return std::abs(elem) >= threshold;
    });
}

template<typename It>
It last_fade_in_sound(It begin, It end) {
    return find_first_above_threshold(begin, above, kFadeInThreshold);
}

template<typename It>
Iterator first_fade_out_sound(Iterator begin, Iterator end) {
    return find_first_above_threshold(begin, above, kFadeOutThreshold);
}

```

Reply:
I understood what you wanted, and to be honest there is nothing in it really, not from the readability point of view (as I see it).  And while having the const and not pushing as much on the stack, should be slightly faster, but not meaningfully faster... and who knows what the the compiler will actually do, so this isn't a big deal. 

However, given that this is the analysis section... is performance important?

Reply:
Not in terms of what this would a difference, no. I'm purely suggesting this because I find it easier to read.

Reply:
How anyone reads lambda functions with closure readable is beyond me anyway, so I have made the changes.  And while I remain slightly concerned about performance, this is not the hill I would choose to die on.

(And besides I find braces on the same line as an 'if' far from readable, as studies into this have shown cf Code Complete 1st ed, so perhaps I am not the best to judge).

Reply:
I'd recommend you to look a bit into functional programming. Learning Haskell is really eye opening ;) I can recommend learnyouahaskell.com

Reply:
I did Prolog at Uni many decades ago.  Functional programming has its value in a few small cases for recursive algorithms and for muti-thread CPUs (which were not readily available when I was studying).  But there are reasons why the vast bulk of code is written in a procedural language.  And, one of those is readability. (Another is performance.)

Reply:
Prolog isn't really a functional language though, its declaritive. If you're coming from the performance angle, I can recommend you to learn rust. Its amazing how nice code can look if its functional-ish while still being blazingly fast. The bulk of code is primarily written because its old and there was nothing better back then thats still performant. That is not the case anymore today. C++ is also a multi-paradigmatic language which includes all the necessary tools to clarify as functional as well. Anyways, it's not very productive to argue about this here. I was just trying to encourage you to broaden your horizon a little bit ;)

Reply:
You are technically correct. As I remember it at the time Prolog was considered to be like a functional language... and the way it works has similarities at least.   Haskal was a thing. There was an elective for Lambda Calculus and Haskal.  Looked too much like math (and formal logic which I did in philosophy) to attract my interest.  I did Networking and Databases.  This was the late 80's early 90's as a mature-age student.

As for Rust, this is one I have already downloaded and considered for my microcontroller projects, mostly in the wake of Counterstrike.  I've backed off a little since then because it doesn't have the lib support yet (for that area).  (And let's face it it is a tad faddish at the moment - yes I am saying that like it is a bad thing - even if Rust will stick.)  

As for my horizons, you do realise that I am in my 60's and that I have been doing this for longer (probably) than you have been around, having first programmed on a Z-80 back in the very early 80's, right?  When I started PC's weren't even a thing.  The Commodore 64 was a year or two away.  And, I had to build my first computer.   Back then you could understand everything that was happening inside the computer.  I nearly built the [EDUC-8](https://en.wikipedia.org/wiki/EDUC-8) back in 1975 when I was still in high school, but could not afford it, and I also doubled the utility of it (I was right - though it would have been a great learning exercise.  And it was complex (for the time).  Switches and lights... no KB.  Back then I was focused on sound and mixers.  And I was more results-focused.  I wanted to mix and sound like the Radio (strangely that hasn't changed!)  In the next seven years, computers went from unobtainable to commonplace - the Commodore 64 I purchased in 1983 was my favourite.)

There is a time in your life when you are young when the "latest and greatest" (= fashionable) always seems to be the best.  And, old fogeys don't seem to adjust well.   However, the 'latest and greatest' doesn't always work out, even if the latter statement is still probably true.  'Amiga' anyone?

You clearly know your stuff, much more than I did at your age.  (I am making assumptions about your age - please correct me if I am wrong).  Since you have taken the liberty of recommending things to me, how about you take a look at Steve McConnell's "Code Complete"?

Reply:
Thank you for your recommendation. I've studied some of Robert C. Martins books so far but "Code Complete" also seems interesting. I will try to get a hold of a copy.

Reply:
Yeah, I've read some Uncle Bob stuff.  I should check out his book(s?).  What I liked about Steve's work was it was so concrete (practical), and backed up by evidence.

So to the matter at hand.  Is there anything else that we need to do to get this merged?

Reply:
I currently don't have the time to give this yet another proper review. Please consult with @daschuer on the next steps.